README.md
# prometheus_exporter_hcloud_sd

prometheus_exporter_hcloud_sd is a CLI tool that discovers Prometheus exporters in your Hetzner Cloud project and generates a Prometeus file-based service discovery YAML configuration for them.

## Description

This tool is particularly useful when running multiple Prometheus exporters in a Hetzner Cloud environment and you want to automate the discovery and configuration process.

`prometheus_exporter_hcloud_sd` utilizes the Hetzner Cloud API for exporter discovery. Therefore, it should be executed within the same network as the servers (it discovers exporters using the servers' private IP).

By default, the tool will discover the following exporters:

- MySQL at port 9104
- Redis at port 9121
- ElasticSearch at port 9114
- Postgres at port 9187
- PgBouncer at port 9127

You can specify additional exporters to discover with the `--exporter` flag, followed by the service name and port, formatted as: `<service_name>=<port>`.

## Installation

Install the gem with:

```bash
gem install prometheus_exporter_hcloud_sd
```

If you use bundler to manage dependencies, install the gem by executing:

```bash
bundle add prometheus_exporter_hcloud_sd
```

## Usage

Before using `prometheus_exporter_hcloud_sd`, ensure the `HCLOUD_READ_TOKEN` environment variable is set with a Hetzner Cloud API token that has read permissions.

```bash
export HCLOUD_READ_TOKEN=your-read-token
```

Run the following command to discover exporters:

```bash
prometheus_exporter_hcloud_sd discover
```

The following is an example of a generated YAML configuration:

```yaml
- targets: ["10.15.1.1:9104"]
  labels:
    job: "mysql"

- targets: ["10.15.1.2:9121","10.15.1.3:9121"]
  labels:
    job: "redis"
```

This YAML configuration can be used to set up the a file-based service discovery in Prometheus. For example, you can create a file called `hcloud_sd.yml` with the above content and then configure Prometheus to use it:

```yaml
scrape_configs:
- job_name: 'node'
  file_sd_configs:
  - files:
    - 'hcloud_sd.yml'
```

To discover additional exporters, use the --exporter flag (can be used multiple times):

```bash
prometheus_exporter_hcloud_sd discover --exporter traefik=80
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at <https://github.com/dunkelbraun/prometheus_exporter_hcloud_sd>.

This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/dunkelbraun/prometheus_exporter_hcloud_sd/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the PrometheusExporterHcloudSd project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/dunkelbraun/prometheus_exporter_hcloud_sd/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.1] - 2023-07-30

### Fixed

- source_code_uri in gemspec.

## [0.1.0] - 2023-07-30

- Initial release


test/test_helper.rb
# frozen_string_literal: true

require "simplecov"
SimpleCov.start do
  enable_coverage :branch
end

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "prometheus_exporter_hcloud_sd"

require "minitest/autorun"
require "mocha/minitest"

require "vcr"
require "climate_control"

VCR.configure do |config|
  config.cassette_library_dir = "test/vcr_cassettes"
  config.filter_sensitive_data("<BEARER_TOKEN>") do |interaction|
    auths = interaction.request.headers["Authorization"].first
    if (match = auths.match(/^Bearer\s+([^,\s]+)/))
      match.captures.first
    end
  end
  config.hook_into :webmock
end

class ThorTestCase < Minitest::Test
  StreamCapture = Struct.new(:stdout, :stderr, :original_stdout, :original_stderr)

  def run_cli_command(*args)
    stream_capture = capture_streams
    begin
      status = 0
      PrometheusExporterHcloudSd::CLI.start(args)
    rescue SystemExit
      status = 1
    end
    rewind_streams && [$stdout.read, $stderr.read, status]
  ensure
    cleanup_streams(stream_capture)
  end

  private

  def capture_streams
    original_stdout = $stdout.dup
    original_stderr = $stderr.dup
    stdout_capture = Tempfile.new("stdout")
    stderr_capture = Tempfile.new("stderr")

    $stdout.reopen(stdout_capture)
    $stderr.reopen(stderr_capture)
    StreamCapture.new(stdout_capture, stderr_capture, original_stdout, original_stderr)
  end

  def cleanup_streams(stream_capture)
    rewind_streams
    $stdout.reopen(stream_capture.original_stdout)
    $stderr.reopen(stream_capture.original_stderr)
    [stream_capture.stdout, stream_capture.stderr].map(&:close)
  end

  def rewind_streams
    [$stdout, $stderr].map(&:rewind)
  end
end


test/test_discover.rb
# frozen_string_literal: true

require "test_helper"
require "open3"

class TestDiscoverCommand < ThorTestCase
  def test_discover_command_with_no_api_authetication
    VCR.use_cassette("hetzner_api_servers_no_auth") do
      ClimateControl.modify(HCLOUD_READ_TOKEN: "some-token") do
        stdout, _stderr, status = run_cli_command("discover")

        assert_equal 0, status

        assert_equal "\n", stdout
      end
    end
  end

  def test_discover_command_with_authetication_but_no_servers
    VCR.use_cassette("hetzner_api_servers_no_servers") do
      ClimateControl.modify(HCLOUD_READ_TOKEN: "some-token") do
        stdout, _stderr, status = run_cli_command("discover")

        assert_equal 0, status

        assert_equal "\n", stdout
      end
    end
  end

  def test_discover_command
    VCR.use_cassette("hetzner_api_servers_auth") do
      mock_service("10.15.1.1", 9104, true)
      mock_service("10.15.1.1", 9121, false)
      mock_service("10.15.1.1", 9114, false)
      mock_service("10.15.1.1", 9187, true)
      mock_service("10.15.1.1", 9127, true)

      mock_service("10.15.1.2", 9104, false)
      mock_service("10.15.1.2", 9121, true)
      mock_service("10.15.1.2", 9114, false)
      mock_service("10.15.1.2", 9187, true)
      mock_service("10.15.1.2", 9127, false)

      mock_service("10.15.1.3", 9104, false)
      mock_service("10.15.1.3", 9121, true)
      mock_service("10.15.1.3", 9114, true)
      mock_service("10.15.1.3", 9187, false)
      mock_service("10.15.1.3", 9127, false)

      ClimateControl.modify(HCLOUD_READ_TOKEN: "some-token") do
        stdout, _stderr, status = run_cli_command("discover")

        assert_equal 0, status

        expected = <<~OUTPUT
          - targets: ["10.15.1.1:9104"]
            labels:
              job: "mysql"
          - targets: ["10.15.1.2:9121","10.15.1.3:9121"]
            labels:
              job: "redis"
          - targets: ["10.15.1.3:9114"]
            labels:
              job: "elastic_search"
          - targets: ["10.15.1.1:9187","10.15.1.2:9187"]
            labels:
              job: "postgres"
          - targets: ["10.15.1.1:9127"]
            labels:
              job: "pgbouncer"
        OUTPUT

        assert_equal expected, stdout
      end
    end
  end

  def test_discover_command_with_additional_exporters
    VCR.use_cassette("hetzner_api_servers_auth") do
      mock_service("10.15.1.1", 9104, true)
      mock_service("10.15.1.1", 9121, false)
      mock_service("10.15.1.1", 9114, false)
      mock_service("10.15.1.1", 9187, true)
      mock_service("10.15.1.1", 9127, true)
      mock_service("10.15.1.1", 80, false)
      mock_service("10.15.1.1", 9177, true)

      mock_service("10.15.1.2", 9104, false)
      mock_service("10.15.1.2", 9121, true)
      mock_service("10.15.1.2", 9114, false)
      mock_service("10.15.1.2", 9187, true)
      mock_service("10.15.1.2", 9127, false)
      mock_service("10.15.1.2", 80, true)
      mock_service("10.15.1.2", 9177, false)

      mock_service("10.15.1.3", 9104, false)
      mock_service("10.15.1.3", 9121, true)
      mock_service("10.15.1.3", 9114, true)
      mock_service("10.15.1.3", 9187, false)
      mock_service("10.15.1.3", 9127, false)
      mock_service("10.15.1.3", 80, false)
      mock_service("10.15.1.3", 9177, false)

      ClimateControl.modify(HCLOUD_READ_TOKEN: "some-token") do
        stdout, _stderr, status = run_cli_command("discover", "--exporter", "traefik=80", "--exporter", "test=9177")

        assert_equal 0, status

        expected = <<~OUTPUT
          - targets: ["10.15.1.1:9104"]
            labels:
              job: "mysql"
          - targets: ["10.15.1.2:9121","10.15.1.3:9121"]
            labels:
              job: "redis"
          - targets: ["10.15.1.3:9114"]
            labels:
              job: "elastic_search"
          - targets: ["10.15.1.1:9187","10.15.1.2:9187"]
            labels:
              job: "postgres"
          - targets: ["10.15.1.1:9127"]
            labels:
              job: "pgbouncer"
          - targets: ["10.15.1.2:80"]
            labels:
              job: "traefik"
          - targets: ["10.15.1.1:9177"]
            labels:
              job: "test"
        OUTPUT

        assert_equal expected, stdout
      end
    end
  end

  def test_discover_command_wthouut_api_token_environment_variable
    _stdout, stderr, status = run_cli_command("discover")

    assert_equal 1, status
    assert_equal "Cannot continue.\nPlease set the HCLOUD_READ_TOKEN environment variable.\n", stderr
  end

  private

  def mock_service(ip_address, port, success)
    Open3.expects(:capture2e).with("timeout 1 nc -zv #{ip_address} #{port}").returns(["", mock(success?: success)])
  end
end


test/test_prometheus_exporter_hcloud_sd.rb
# frozen_string_literal: true

require "test_helper"

class TestPrometheusExporterHcloudSd < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::PrometheusExporterHcloudSd::VERSION
  end
end


lib/prometheus_exporter_hcloud_sd/discovery_config.rb
require "json"
require "open3"
require_relative "hetzner_servers"

module PrometheusExporterHcloudSd
  class DiscoveryConfig
    PORTS = {
      "9104" => :mysql,
      "9121" => :redis,
      "9114" => :elastic_search,
      "9187" => :postgres,
      "9127" => :pgbouncer
    }.freeze

    def self.generate(options)
      new(options).generate
    end

    def initialize(options)
      @additional_ports = options[:exporter].each_with_object({}) do |exporter, acc|
        service, port = exporter.split("=")
        acc[port] = service.to_sym
      end
    end

    def generate
      server_ips = PrometheusExporterHcloudSd::HetznerServers.private_ips
      output = ""

      PORTS.merge(@additional_ports).each do |port, service|
        available_servers = discover_exporters(server_ips: server_ips, port: port)

        next if available_servers.none?

        output << target_configuration(service: service, available_servers: available_servers)
      end
      output
    end

    def discover_exporters(server_ips:, port:)
      server_ips.collect do |ip|
        command = "timeout 1 nc -zv #{ip} #{port}"
        _, status = Open3.capture2e(command)
        status.success? ? "#{ip}:#{port}" : nil
      end.compact
    end

    def target_configuration(service:, available_servers:)
      <<~LINE
        - targets: #{available_servers.to_json}
          labels:
            job: "#{service}"
      LINE
    end
  end
end


lib/prometheus_exporter_hcloud_sd/recipes/discover.rb
require "json"
require "net/http"
require "uri"
require "open3"

SERVERS_URI = URI.parse("https://api.hetzner.cloud/v1/servers").freeze
token = begin
  ENV.fetch("HCLOUD_READ_TOKEN")
rescue KeyError
  raise Thor::Error, "Cannot continue.\nPlease set the HCLOUD_READ_TOKEN environment variable."
end

AUTH_HEADER = { Authorization: "Bearer #{token}" }.freeze

response = Net::HTTP.get(SERVERS_URI, AUTH_HEADER)
json_response = JSON.parse(response, symbolize_names: true)

server_ips = json_response.fetch(:servers, []).map do |server|
  server.fetch(:private_net, [{}]).first.fetch(:ip, nil)
end.compact

PORTS = {
  "9104" => :mysql,
  "9121" => :redis,
  "9114" => :elastic_search,
  "9187" => :postgres,
  "9127" => :pgbouncer
}.freeze

output = ""

additional_ports = options[:exporter].each_with_object({}) do |exporter, acc|
  service, port = exporter.split("=")
  acc[port] = service.to_sym
end

PORTS.merge(additional_ports).each do |port, service|
  available_servers = server_ips.collect do |ip|
    command = "timeout 1 nc -zv #{ip} #{port}"
    _, status = Open3.capture2e(command)
    status.success? ? "#{ip}:#{port}" : nil
  end.compact

  next if available_servers.none?

  output << <<~LINE
    - targets: #{available_servers.to_json}
      labels:
        job: "#{service}"
  LINE
end

puts output


lib/prometheus_exporter_hcloud_sd.rb
# frozen_string_literal: true

require "thor"
require_relative "prometheus_exporter_hcloud_sd/version"
require_relative "prometheus_exporter_hcloud_sd/discovery_config"

module PrometheusExporterHcloudSd
  class Error < StandardError; end

  class CLI < Thor
    def self.exit_on_failure?
      true
    end

    desc "discover", "Discover Prometheus exporters in your Hetzner Cloud project."
    long_desc <<~LONGDESC
      `prometheus_exporter_hcloud_sd discover` will print out a YAML configuration for discovered Prometheus exporters
      in your Hetzner Cloud project using the Hetzner Cloud API.

      To authenticate with the API, set the `HCLOUD_READ_TOKEN` environment variable with an API token that has read
      permissions.

      Needs to be run inside the same network as the servers (uses the private IP of the servers).

      By default it will discover the following exporters

        - mysql at port 9104

        - redis at port 9121

        - elastic_search at port 9114

        - postgres at port 9187

        - pgbounce at port 9127

      Example:

      > $ HCLOUD_READ_TOKEN=your-read-token prometheus_exporter_hcloud_sd discover

      - targets: ["10.15.1.1:9104"]

          labels:

            job: "mysql"

      - targets: ["10.15.1.2:9121","10.15.1.3:9121"]

        labels:

          job: "redis"
    LONGDESC
    option :exporter, type: :string, repeatable: true, default: [],
                      desc: "Additional exporters to discover. Format: <service_name>=<port>"
    def discover
      puts PrometheusExporterHcloudSd::DiscoveryConfig.generate(options)
    end
  end
end
